### PR TITLE
build: add nconfig

### PR DESCRIPTION
**Pull request by 532910**
Add support for make target nconfig (ncurses)

Signed-off-by: Sergio E. Nemirowski <sergio@outerface.net>

### CODE REVIEW

**Comment by ynezz** [comment]
@cotequeiroz can you check this please?

**Comment by aparcar** [comment]
Tested and works, seems to use more space of the screen and the search function is better. Cool.

File: scripts/config/nconf.c [hunk 1]
```diff
@@ -0,0 +1,1566 @@
...
+"\n"
+"o  Local help:  To get help related to the current menu entry, use any\n"
+"   of <?> <h>, or if in a data entry window then press <F1>.\n"
+"\n"
+"\n"
+"Menu entries\n"
+"------------\n"
+"This interface lets you select features and parameters for the kernel\n"
+"build.  Kernel features can either be built-in, modularized, or removed.\n"
+"Parameters must be entered as text or decimal or hexadecimal numbers.\n"
```
Comment (cotequeiroz) [inline scripts/config/nconf.c:34]: I suggest changing the "kernel" references in the help message to "OpenWrt".  You may use the text in `mconf.c` if you wish.

**Review by cotequeiroz — CHANGES_REQUESTED** [review]
This looks good.  I just have some nitpicks that don't affect usability:
- keep the nconf block in the Makefile exactly as upstream
- I would change the "kernel" references in the help message in nconf.c.
- update the README with the changes we make from upstream.  Since this is actually reverting a change we made, just removing the `nconf` word from the first item of the list is enough.  This should do the job:
```
sed -i -e 's/nconf, //' scripts/config/README
```

File: scripts/config/nconf.c [hunk 2]
```diff
@@ -0,0 +1,1569 @@
...
+"   you the global help window, which you are just reading.\n"
+"\n"
+"o  A short version of the global help is available by pressing <F3>.\n"
+"\n"
+"o  Local help:  To get help related to the current menu entry, use any\n"
+"   of <?> <h>, or if in a data entry window then press <F1>.\n"
+"\n"
+"\n"
+"OpenWrt config is based on Kernel kconfig\n"
+"so ipkg packages are referred here as modules.\n"
```
Comment (532910) [inline scripts/config/nconf.c:31]: What about to add these 2 lines to mconf.c

File: scripts/config/Makefile [hunk 3]
```diff
@@ -64,6 +64,16 @@ $(foreach f, mconf.o $(lxdialog), \
...
+# nconf: Used for the nconfig target based on ncurses
+hostprogs-y	+= nconf
+nconf-objs	:= nconf.o nconf.gui.o $(common-objs)
+
+HOSTLDLIBS_nconf = $(shell . $(obj)/nconf-cfg && echo $$libs)
+HOSTCFLAGS_nconf.o = $(shell . $(obj)/nconf-cfg && echo $$cflags))
+
+$(obj)/nconf.o: $(obj)/nconf-cfg
+$(obj)/nconf.gui.o: $(obj)/nconf-cfg
+
```
[the remark this answers is not in the document]
Comment (532910) [reply scripts/config/Makefile]: Is `HOSTCFLAGS_nconf.gui.o` not superfluous? But I'm agree to keep it as close to upstream as possible

File: scripts/config/nconf.c (hunk 1 above)
Comment (532910) [reply scripts/config/nconf.c]: Comparing mconf.c and nconf.c from the upstream, the Linux` word is removed from mconf.c:

```
"This interface lets you select features and parameters for the build.\n"
"Features can either be built-in, modularized, or ignored. Parameters\n"
"must be entered in as decimal or hexadecimal numbers or text.\n"
```

Comment (cotequeiroz) [reply scripts/config/nconf.c]: I meant openwrt mconf.c, not upstream.  I suggested the text from mconf.c for consistency, but your suggestion below can work.

**Comment by cotequeiroz** [comment]
I'm sorry for not being clear enough about the `Makefile` changes.

I'm keeping track of changes we make from upstram kconfig here: https://github.com/cotequeiroz/linux.  You can see my commits with the changes needed to get from upstream to openwrt here: https://github.com/cotequeiroz/linux/commits/openwrt

So when we need to upgrade `config` to a newer upstream version, we know what changes were needed to get there.

The intention is to keep these changes to a minimum.  So the `Makefile` portion you're adding here in the PR should **exactly** match the original upstream file, including white space and location.

Here's a patch you may apply to the current state of this PR:
``` diff
--- a/scripts/config/Makefile
+++ b/scripts/config/Makefile
@@ -52,6 +52,16 @@ HOSTCFLAGS_parser.tab.o	:= -I $(srctree)/$(src)
 hostprogs-y	+= conf
 conf-objs	:= conf.o $(common-objs)
 
+# nconf: Used for the nconfig target based on ncurses
+hostprogs-y	+= nconf
+nconf-objs	:= nconf.o nconf.gui.o $(common-objs)
+
+HOSTLDLIBS_nconf	= $(shell . $(obj)/nconf-cfg && echo $$libs)
+HOSTCFLAGS_nconf.o	= $(shell . $(obj)/nconf-cfg && echo $$cflags)
+HOSTCFLAGS_nconf.gui.o	= $(shell . $(obj)/nconf-cfg && echo $$cflags)
+
+$(obj)/nconf.o $(obj)/nconf.gui.o: $(obj)/nconf-cfg
+
 # mconf: Used for the menuconfig target based on lxdialog
 hostprogs-y	+= mconf
 lxdialog	:= $(addprefix lxdialog/, \
@@ -64,15 +74,6 @@ $(foreach f, mconf.o $(lxdialog), \
 
 $(addprefix $(obj)/, mconf.o $(lxdialog)): $(obj)/mconf-cfg
 
-# nconf: Used for the nconfig target based on ncurses
-hostprogs-y	+= nconf
-nconf-objs	:= nconf.o nconf.gui.o $(common-objs)
-
-HOSTLDLIBS_nconf = $(shell . $(obj)/nconf-cfg && echo $$libs)
-HOSTCFLAGS_nconf.o = $(shell . $(obj)/nconf-cfg && echo $$cflags))
-
-$(obj)/nconf.o $(obj)/nconf.gui.o: $(obj)/nconf-cfg
-
 # qconf: Used for the xconfig target based on Qt
 hostprogs-y	+= qconf
 qconf-cxxobjs	:= qconf.o
@@ -80,7 +81,6 @@ qconf-objs	:= images.o $(common-objs)
 
 HOSTLDLIBS_qconf	= $(shell . $(obj)/qconf-cfg && echo $$libs)
 HOSTCXXFLAGS_qconf.o	= $(shell . $(obj)/qconf-cfg && echo $$cflags)
-HOSTCFLAGS_nconf.gui.o	= $(shell . $(obj)/nconf-cfg && echo $$cflags)
 
 $(obj)/qconf.o: $(obj)/qconf-cfg $(obj)/qconf.moc
 
```

**Comment by 532910** [comment]
Makefile is correctly updated now.

**Review by 532910 — COMMENTED** [review]
> I meant openwrt mconf.c, not upstream. I suggested the text from mconf.c for consistency, but your suggestion below can work.

Yep, it was clear. But I meant the same intention to keep changes to a minimum (:
Instead of rewriting  the first sentence, just add one more before.

But this only makes sense in junction with  #3112

**Review by cotequeiroz — APPROVED** [review]
LGTM

`Reviewed-by: Eneas U de Queiroz <cotequeiroz@gmail.com>`

**Comment by 532910** [comment]
Who should approve this?

**Comment by ynezz** [comment]
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

**Comment by 532910** [comment]
This PR is only valid with #3112, otherwise comment should be updated.

**Pull request closed without merging**